### ISSUE TITLE
WebFlutterDriver.requestData command doesn't work with JSON string input

### ISSUE THREAD

**Issue by voobel** [flutter/flutter#71267]
## Steps to Reproduce
1. Add logging (`print(message);`) to `flutter/packages/flutter_driver/lib/src/extension/_extension_web.dart#registerWebServiceExtension` so you can monitor incoming flutter_driver commands in browser console.
2. In flutter_driver **(web)** test use `await driver.requestData(jsonEncode(command));` where command is a map. It will send stringified JSON command to app.

**Expected results:**
Request JSON string is escaped correctly in `request_data` command `message` and can be parsed successfully.

**Actual results:** 
Request JSON is broken (double quoation marks not escaped), for example: `{"command":"request_data","message":"{"command":"setDataItem","objectId":"Element223","value":2}"}`. It should be like this instead: `{"command":"request_data","message":"{\"command\":\"setDataItem\",\"objectId\":\"Element223\",\"value\":2}"}`.
Exception such as `FormatException: SyntaxError: Unexpected token c in JSON at position 39` is thrown in `_extension_web.dart:25`

## Additional information
Exactly same (JSON encoded) `request_data` messages are working with mobile FlutterDriver.

## Work-around
Do replacements explicitly if using WebFlutterDriver:
```
String _prepareJsonCommand(Map<String, dynamic> command) {
  if (Platform.environment['FLUTTER_WEB_TEST'] != null) {
    return jsonEncode(command).replaceAll('"', '\\"');
  }
  return jsonEncode(command);
}
```

<details>
  <summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel unknown, 1.22.2, on Mac OS X 10.15.6 19G2021, locale en-EE)
    • Flutter version 1.22.2 at /Users/siim/git/flutter
    • Framework revision 84f3d28555 (6 weeks ago), 2020-10-15 16:26:19 -0700
    • Engine revision b8752bbfff
    • Dart version 2.10.2

 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /Users/siim/Library/Android/sdk
    • Platform android-30, build-tools 29.0.2
    • ANDROID_HOME = /Users/siim/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

 
[✓] Xcode - develop for iOS and macOS (Xcode 12.0.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.0.1, Build version 12A7300
    • CocoaPods version 1.9.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[!] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)

[!] IntelliJ IDEA Community Edition (version 2020.2.3)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • For information about installing plugins, see
      https://flutter.dev/intellij-setup/#installing-the-plugins

[✓] Connected device (2 available)
    • Web Server (web) • web-server • web-javascript • Flutter Tools
    • Chrome (web)     • chrome     • web-javascript • Google Chrome 86.0.4240.198
    ! Error: Siim’s iPhone is not connected. Xcode will continue when Siim’s iPhone is connected. (code -13)

! Doctor found issues in 2 categories.
```

</details>

**Comment by darshankawar** [comment]
@voobel 
Can you provide more details on how have you setup `webFlutterDriver` ? Are [these](https://github.com/flutter/flutter/issues/49843#issuecomment-585372148) the steps to set it up ?

**Comment by voobel** [comment]
I have application running locally: `flutter run --web-port=8081 --target=lib/automation/main_web.dart`
Then from Python side I call Dart tests with `subprocess.Popen` like: `{flutter_cmd} packages pub run test --tags {self.test_tags} {self.testcase_file} --reporter json`
There are also some things done in Python side beforehand, like:
```python
os.environ['DRIVER_SESSION_ID'] = driver.session_id
os.environ['DRIVER_SESSION_URI'] = f'http://127.0.0.1:{port}/'
os.environ['DRIVER_SESSION_SPEC'] = 'webdriverspec.w3c'
os.environ['DRIVER_SESSION_CAPABILITIES'] = json.dumps(driver.capabilities)
```
so that Dart tests know how to connect:
```dart
/// Initializes [WebFlutterDriver] in common way.
///
/// Bits and pieces are stolen from [WebFlutterDriver.connectWeb] and [FlutterWebConnection.connect].
/// [WebFlutterDriver.connectWeb] can't be used directly because it includes direct URL navigation,
/// which in our case is not always wanted.
Future<FlutterDriver> initWebDriver({Duration timeout = appConnectTimeout}) async {
  final Map<String, dynamic> settings = <String, dynamic>{
    'support-timeline-action': Platform.environment['SUPPORT_TIMELINE_ACTION'] == 'true',
    'session-id': Platform.environment['DRIVER_SESSION_ID'],
    'session-uri': Platform.environment['DRIVER_SESSION_URI'],
    'session-spec': Platform.environment['DRIVER_SESSION_SPEC'],
    'android-chrome-on-emulator': Platform.environment['ANDROID_CHROME_ON_EMULATOR'] == 'true',
    'session-capabilities': Platform.environment['DRIVER_SESSION_CAPABILITIES'],
  };
  final String sessionId = settings['session-id'].toString();
  final Uri sessionUri = Uri.parse(settings['session-uri'].toString());
  final async_io.WebDriver driver = async_io.WebDriver(
      sessionUri,
      sessionId,
      json.decode(settings['session-capabilities'] as String) as Map<String, dynamic>,
      async_io.AsyncIoRequestClient(sessionUri.resolve('session/$sessionId/')),
      _convertToSpec(settings['session-spec'].toString().toLowerCase()));
  await waitUntilExtensionInstalled(driver, timeout);
  final FlutterWebConnection connection = FlutterWebConnection(driver, settings['support-timeline-action'] as bool);
  return WebFlutterDriver.connectedTo(connection);
}
```
If you're asking why not `flutter drive`, then there are multiple reasons for this:
* We don't have access to app codebase from tests for security reasons. `flutter drive` allows to run only from app codebase.
* In Python side we're doing many different "setup steps" beforehand, such as test data generating with API-s & also using Selenium browser to reach point where Flutter Web app is embedded in an iframe & then "context switch" to Dart side.

Long story short -- I doubt the way it's executed is causing problems. Web automation in general is working fine.

**Comment by yjbanov** [comment]
Driver requests are encoded as URL parameters. They were not built for sending arbitrary data, so this might be a new feature we'd have to add.

**Comment by arturk** [comment]
It becomes even more difficult when you try to encode the command from the browser console. I've created a chrome extension to debug my driver commands here:
https://github.com/arturk/flutter-driver-chrome-extension
In case of JSON payload I have to encode json quotes in my extension like `var magicQuote = '\\\\\\\\\\\\"';`
